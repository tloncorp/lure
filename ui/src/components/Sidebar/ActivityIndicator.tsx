import cn from 'classnames';
import React from 'react';
import BulletIcon from '../icons/BulletIcon';

interface ActivityIndicatorProps {
  count: number;
  bg?: string;
  className?: string;
}

export default function ActivityIndicator({
  count,
  bg = 'bg-gray-100',
  className,
}: ActivityIndicatorProps) {
  return (
    <div
      className={cn(
        'flex h-6 w-6 items-center justify-center rounded text-sm font-semibold',
        bg,
        className
      )}
    >
      {count === 0 ? (
        <BulletIcon className="h-6 w-6" />
      ) : count > 99 ? (
        '99+'
      ) : (
        count
      )}
    </div>
  );
}
