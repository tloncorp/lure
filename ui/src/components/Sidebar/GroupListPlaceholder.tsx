import * as React from 'react';
import cn from 'classnames';
import { randomElement, randomIntInRange } from '@/logic/utils';

function GroupItemPlaceholder() {
  const background = `bg-gray-${randomElement([100, 200, 400])}`;

  return (
    <li className="relative flex w-full items-center justify-between rounded-lg">
      <div className="flex w-full flex-1 items-center space-x-3 rounded-lg p-2">
        <div className="flex h-12 w-12 items-center justify-center rounded-md bg-gray-100 text-sm sm:h-6 sm:w-6" />
        <div
          className={cn(background, 'h-4 rounded-md')}
          style={{ width: `${randomIntInRange(50, 180)}px` }}
        />
      </div>
    </li>
  );
}

interface GroupListPlaceholderProps {
  count: number;
}

function GroupListPlaceholder({ count }: GroupListPlaceholderProps) {
  return (
    <ul className="h-full animate-pulse space-y-3 p-2 sm:space-y-0">
      {new Array(count).fill(true).map((_, i) => (
        <GroupItemPlaceholder key={i} />
      ))}
    </ul>
  );
}

export default React.memo(GroupListPlaceholder);
