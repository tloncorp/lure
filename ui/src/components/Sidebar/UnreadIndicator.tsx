import React from 'react';
import cn from 'classnames';

import ActivityIndicator from './ActivityIndicator';

export default function UnreadIndicator({ className }: { className?: string }) {
  return (
    <ActivityIndicator
      count={0}
      bg={'transparent'}
      className={cn('text-blue', className)}
    />
  );
}
