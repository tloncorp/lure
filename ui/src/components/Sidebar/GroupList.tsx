import cn from 'classnames';
import React, { PropsWithChildren, useEffect, useMemo } from 'react';
import { uniq, without } from 'lodash';
import { DndProvider, useDrag, useDrop } from 'react-dnd';
import { HTML5Backend } from 'react-dnd-html5-backend';
import { TouchBackend } from 'react-dnd-touch-backend';
import * as Popover from '@radix-ui/react-popover';
import { Virtuoso } from 'react-virtuoso';
import { useIsMobile } from '@/logic/useMedia';
import {
  useGang,
  useGangList,
  useGroup,
  useGroupsInitialized,
  useGroupState,
} from '@/state/groups/groups';
import { SettingsState, useSettingsState } from '@/state/settings';
import GroupAvatar from '@/groups/GroupAvatar';
import GroupActions from '@/groups/GroupActions';
import { Group } from '@/types/groups';
import { getFlagParts } from '@/logic/utils';
import { useHasMigratedChannels } from '@/logic/useMigrationInfo';
import SidebarItem from './SidebarItem';
import GroupListPlaceholder from './GroupListPlaceholder';
import Bullet16Icon from '../icons/Bullet16Icon';
import MigrationTooltip from '../MigrationTooltip';

const dragTypes = {
  GROUP: 'group',
};

const selOrderedPins = (s: SettingsState) => ({
  order: s.groups.orderedGroupPins,
  loaded: s.loaded,
});

function DraggableGroupItem({ flag }: { flag: string }) {
  const group = useGroup(flag);

  const [{ isDragging }, drag] = useDrag(() => ({
    type: dragTypes.GROUP,
    item: { flag },
    collect: (monitor) => ({
      isDragging: !!monitor.isDragging(),
    }),
  }));

  return (
    <div
      ref={drag}
      className={cn(
        'absolute w-full',
        isDragging ? 'opacity-0' : 'opacity-100'
      )}
    >
      <SidebarItem
        div
        icon={<GroupAvatar size="h-12 w-12 sm:h-6 sm:w-6" {...group?.meta} />}
        actions={<GroupActions flag={flag} />}
        to={`/groups/${flag}`}
      >
        {group?.meta.title}
      </SidebarItem>
    </div>
  );
}

const GroupItem = React.memo(({ flag }: { flag: string }) => {
  const group = useGroup(flag);
  const { ship } = getFlagParts(flag);
  const isMigrated = useHasMigratedChannels(flag);

  if (!isMigrated) {
    return (
      <MigrationTooltip ship={ship} flag={flag} side="right" kind="group">
        <SidebarItem
          className="opacity-60"
          icon={<GroupAvatar size="h-12 w-12 sm:h-6 sm:w-6" {...group?.meta} />}
          actions={
            <Bullet16Icon
              className="m-2 h-4 w-4 text-orange opacity-60"
              aria-label="Pending Migration"
            />
          }
        >
          {group?.meta.title}
        </SidebarItem>
      </MigrationTooltip>
    );
  }

  return (
    <SidebarItem
      icon={<GroupAvatar size="h-12 w-12 sm:h-6 sm:w-6" {...group?.meta} />}
      actions={<GroupActions flag={flag} />}
      to={`/groups/${flag}`}
    >
      {group?.meta.title}
    </SidebarItem>
  );
});

function GroupItemContainer({
  flag,
  children,
}: PropsWithChildren<{ flag: string }>) {
  const { order } = useSettingsState(selOrderedPins);
  const [{ isOver }, drop] = useDrop<
    { flag: string },
    undefined,
    { isOver: boolean }
  >(
    () => ({
      accept: dragTypes.GROUP,
      drop: ({ flag: itemFlag }) => {
        if (!itemFlag || itemFlag === flag) {
          return undefined;
        }
        // [1, 2, 3, 4] 1 -> 3
        // [2, 3, 4]
        const beforeSlot = order.indexOf(itemFlag) < order.indexOf(flag);
        const orderWithoutOriginal = without(order, itemFlag);
        const slicePoint = orderWithoutOriginal.indexOf(flag);
        // [2, 3] [4]
        const left = orderWithoutOriginal.slice(
          0,
          beforeSlot ? slicePoint + 1 : slicePoint
        );
        const right = orderWithoutOriginal.slice(slicePoint);
        // concat([2, 3], [1], [4])
        const newOrder = uniq(left.concat([itemFlag], right));
        // [2, 3, 1, 4]
        useSettingsState
          .getState()
          .putEntry('groups', 'orderedGroupPins', newOrder);

        return undefined;
      },
      collect: (monitor) => ({
        isOver: !!monitor.isOver(),
      }),
    }),
    [flag]
  );

  return (
    <li
      ref={drop}
      className={cn(
        'relative flex h-16 w-full ring-4 sm:h-10',
        isOver && 'ring-blue-500',
        !isOver && 'ring-transparent'
      )}
    >
      {children}
    </li>
  );
}

// Gang is a pending group invite
function GangItem(props: { flag: string }) {
  const { flag } = props;
  const { preview, claim } = useGang(flag);
  const isMobile = useIsMobile();

  if (!claim) {
    return null;
  }

  const requested = claim.progress === 'knocking';
  const errored = claim.progress === 'error';
  const handleCancel = async () => {
    if (requested) {
      await useGroupState.getState().rescind(flag);
    } else {
      await useGroupState.getState().cancel(flag);
    }
  };

  return (
    <Popover.Root>
      <Popover.Anchor>
        <Popover.Trigger asChild>
          <SidebarItem
            icon={
              <GroupAvatar
                {...preview?.meta}
                size="h-12 w-12 sm:h-6 sm:w-6"
                className="opacity-60"
              />
            }
          >
            <span className="inline-block w-full truncate opacity-60">
              {preview ? preview.meta.title : flag}
            </span>
          </SidebarItem>
        </Popover.Trigger>
      </Popover.Anchor>
      <Popover.Content
        side={isMobile ? 'top' : 'right'}
        sideOffset={isMobile ? 0 : 16}
        className="z-10"
      >
        <div className="flex w-[200px] flex-col space-y-4 rounded-lg bg-white p-4 leading-5 drop-shadow-lg">
          {requested ? (
            <>
              <span>You've requested to join this group.</span>
              <span>
                An admin will have to approve your request and then you'll
                receive an invitation to join.
              </span>
            </>
          ) : errored ? (
            <>
              <span>You were unable to join the group.</span>
              <span>
                The group may not exist or they may be running an incompatible
                version.
              </span>
            </>
          ) : (
            <>
              <span>You are currently joining this group.</span>
              <span>
                It may take a few minutes depending on the host&apos;s and your
                connection.
              </span>
            </>
          )}
          {(errored || requested) && (
            <div className="flex">
              <Popover.Close>
                <button
                  className="small-button bg-gray-50 text-gray-800"
                  onClick={handleCancel}
                >
                  {requested ? 'Cancel Request' : 'Cancel Join'}
                </button>
              </Popover.Close>
            </div>
          )}
        </div>
      </Popover.Content>
    </Popover.Root>
  );
}

function itemContent(_i: number, [flag, _group]: [string, Group]) {
  return <GroupItem key={flag} flag={flag} />;
}

interface GroupListProps {
  className?: string;
  pinned?: boolean;
  groups: [string, Group][];
  pinnedGroups: [string, Group][];
}

export default function GroupList({
  className,
  pinned = false,
  groups,
  pinnedGroups,
}: GroupListProps) {
  const isMobile = useIsMobile();
  const gangs = useGangList();
  const initialized = useGroupsInitialized();
  const { order, loaded } = useSettingsState(selOrderedPins);
  const thresholds = {
    atBottomThreshold: 125,
    atTopThreshold: 125,
    overscan: isMobile
      ? { main: 200, reverse: 200 }
      : { main: 400, reverse: 400 },
  };

  const groupsWithoutPinned = useMemo(
    () =>
      groups.filter(
        ([flag, _g]) => !pinnedGroups.map(([f, _]) => f).includes(flag)
      ),
    [groups, pinnedGroups]
  );

  useEffect(() => {
    const hasKeys = order && !!order.length;
    const pinnedKeys = Object.keys(pinnedGroups);
    const hasPinnedKeys = pinnedKeys.length > 0;

    if (!loaded) {
      return;
    }

    // Correct order state, fill if none, remove duplicates, and remove
    // old uninstalled app keys
    if (!hasKeys && hasPinnedKeys) {
      useSettingsState
        .getState()
        .putEntry('groups', 'orderedGroupPins', pinnedKeys);
    } else if (order.length < pinnedKeys.length) {
      useSettingsState
        .getState()
        .putEntry('groups', 'orderedGroupPins', uniq(order.concat(pinnedKeys)));
    } else if (order.length > pinnedKeys.length && hasPinnedKeys) {
      useSettingsState
        .getState()
        .putEntry(
          'groups',
          'orderedGroupPins',
          uniq(order.filter((key) => key in pinnedGroups).concat(pinnedKeys))
        );
    }
  }, [pinnedGroups, order, loaded]);

  const pinnedGroupsOptions = useMemo(
    () =>
      pinnedGroups.map(([flag]) => (
        <GroupItemContainer flag={flag} key={flag}>
          <DraggableGroupItem flag={flag} />
        </GroupItemContainer>
      )),
    [pinnedGroups]
  );

  if (!initialized) {
    if (pinned) return null;

    return <GroupListPlaceholder count={groups.length || 5} />;
  }

  return pinned ? (
    <DndProvider
      backend={isMobile ? TouchBackend : HTML5Backend}
      options={
        isMobile
          ? {
              delay: 50,
              scrollAngleRanges: [
                { start: 30, end: 150 },
                { start: 210, end: 330 },
              ],
            }
          : undefined
      }
    >
      <li className="-mx-2 mt-5 grow border-t-2 border-gray-50 pt-3 pb-2">
        <span className="ml-4 text-sm font-semibold text-gray-400">
          Pinned Groups
        </span>
      </li>
      {pinnedGroupsOptions}
    </DndProvider>
  ) : (
    <>
      {gangs.map((flag) => (
        <GangItem key={flag} flag={flag} />
      ))}
      <Virtuoso
        {...thresholds}
        data={groupsWithoutPinned}
        computeItemKey={(_i, [flag]) => flag}
        itemContent={itemContent}
        className="h-full w-full overflow-x-hidden"
      />
    </>
  );
}
