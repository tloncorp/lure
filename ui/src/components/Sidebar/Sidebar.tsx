import React, { useState, useRef, useMemo, useCallback } from 'react';
import cn from 'classnames';
import { useLocation } from 'react-router';
import ActivityIndicator from '@/components/Sidebar/ActivityIndicator';
import MobileSidebar from '@/components/Sidebar/MobileSidebar';
import GroupList from '@/components/Sidebar/GroupList';
import * as DropdownMenu from '@radix-ui/react-dropdown-menu';
import { useGroups, usePendingInvites } from '@/state/groups';
import { useIsMobile } from '@/logic/useMedia';
import AppGroupsIcon from '@/components/icons/AppGroupsIcon';
import MagnifyingGlass from '@/components/icons/MagnifyingGlass16Icon';
import SidebarItem from '@/components/Sidebar/SidebarItem';
import AddIcon16 from '@/components/icons/Add16Icon';
import SidebarSorter from '@/components/Sidebar/SidebarSorter';
import { usePinnedGroups } from '@/state/chat';
import { hasKeys } from '@/logic/utils';
import ShipName from '@/components/ShipName';
import Avatar, { useProfileColor } from '@/components/Avatar';
import useGroupSort from '@/logic/useGroupSort';
import { useNotifications } from '@/notifications/useNotifications';
import { debounce } from 'lodash';
import ArrowNWIcon from '../icons/ArrowNWIcon';
import MenuIcon from '../icons/MenuIcon';

export function GroupsAppMenu() {
  const [menuOpen, setMenuOpen] = useState(false);
  return (
    <SidebarItem
      div
      className={cn(
        menuOpen
          ? 'bg-gray-100 text-gray-800'
          : 'text-black hover:text-gray-800',
        'group'
      )}
      icon={
        <DropdownMenu.Root onOpenChange={() => setMenuOpen(!menuOpen)}>
          <DropdownMenu.Trigger asChild className="appearance-none">
            <div className={cn('h-6 w-6 rounded group-hover:bg-gray-100')}>
              <AppGroupsIcon
                className={cn(
                  'h-6 w-6',
                  menuOpen ? 'hidden' : 'group-hover:hidden'
                )}
              />
              <MenuIcon
                aria-label="Open Menu"
                className={cn(
                  'm-1 h-4 w-4 text-gray-800',
                  menuOpen ? 'block' : 'hidden group-hover:block'
                )}
              />
            </div>
          </DropdownMenu.Trigger>

          <DropdownMenu.Content className="dropdown mt-2 -ml-2 w-60">
            <a
              className="no-underline"
              href="https://airtable.com/shrflFkf5UyDFKhmW"
              target="_blank"
              rel="noreferrer"
            >
              <DropdownMenu.Item className="dropdown-item pl-3 text-blue">
                Submit Feedback
              </DropdownMenu.Item>
            </a>
          </DropdownMenu.Content>
        </DropdownMenu.Root>
      }
    >
      <div className="flex items-center justify-between">
        Groups
        <a
          title="Back to Landscape"
          aria-label="Back to Landscape"
          href="/apps/grid"
          target="_blank"
          rel="noreferrer"
          className={cn(
            'h-6 w-6 no-underline',
            menuOpen ? 'block' : 'hidden group-hover:block'
          )}
        >
          <ArrowNWIcon className="text-gray-400" />
        </a>
      </div>
    </SidebarItem>
  );
}

export default function Sidebar() {
  const isMobile = useIsMobile();
  const location = useLocation();
  const pendingInvites = usePendingInvites();
  const [scrolledFromTop, setScrolledFromTop] = useState(false);

  const pendingInvitesCount = pendingInvites.length;
  const { count } = useNotifications();

  const { sortFn, setSortFn, sortOptions, sortGroups } = useGroupSort();
  const groups = useGroups();
  const pinnedGroups = usePinnedGroups();
  const sortedGroups = sortGroups(groups);
  const sortedPinnedGroups = sortGroups(pinnedGroups);
  const shipColor = useProfileColor(window.our);

  const ref = useRef<HTMLUListElement>(null);

  const classes = useMemo(
    () =>
      cn(
        'flex w-full flex-col space-y-1 px-2 pt-2',
        scrolledFromTop && 'bottom-shadow'
      ),
    [scrolledFromTop]
  );

  // to prevent re-render, only set state when necessary
  const scrollHandler = useCallback(() => {
    debounce(() => {
      if (ref.current?.scrollTop === 0) {
        if (!scrolledFromTop) {
          setScrolledFromTop(true);
        }
      } else {
        // eslint-disable-next-line no-lonely-if
        if (scrolledFromTop) {
          setScrolledFromTop(false);
        }
      }
    }, 100);
  }, [scrolledFromTop]);

  if (isMobile) {
    return <MobileSidebar />;
  }

  return (
    <nav className="flex h-full w-64 flex-col bg-white">
      <ul className={classes}>
        {/* TODO: FETCH WINDOW.OUR WITHOUT IT RETURNING UNDEFINED */}
        <GroupsAppMenu />
        <div className="h-5" />
        <SidebarItem
          highlight={shipColor}
          icon={<Avatar size="xs" ship={window.our} />}
          to={'/profile/edit'}
        >
          <ShipName showAlias name={window.our} />
        </SidebarItem>
        <SidebarItem
          icon={<ActivityIndicator count={count} />}
          to={`/notifications`}
          defaultRoute
        >
          Notifications
        </SidebarItem>
        <SidebarItem
          icon={<MagnifyingGlass className="m-1 h-4 w-4" />}
          to="/find"
        >
          <div className="flex items-center">
            Find Groups
            {pendingInvitesCount > 0 ? (
              <span className="ml-auto font-semibold text-blue">
                {pendingInvitesCount}
              </span>
            ) : null}
          </div>
        </SidebarItem>
        <SidebarItem
          icon={<AddIcon16 className="m-1 h-4 w-4" />}
          to="/groups/new"
          state={{ backgroundLocation: location }}
        >
          Create Group
        </SidebarItem>
      </ul>
      <ul
        ref={ref}
        onScroll={scrollHandler}
        className="flex-initial overflow-y-auto overflow-x-hidden px-2"
      >
        {hasKeys(pinnedGroups) ? (
          <GroupList
            className="p-2"
            pinned
            groups={sortedGroups}
            pinnedGroups={sortedPinnedGroups}
          />
        ) : null}
        <li className="-mx-2 mt-5 grow border-t-2 border-gray-50 pt-3 pb-2">
          <span className="ml-4 text-sm font-semibold text-gray-400">
            All Groups
          </span>
        </li>
        <li className="relative p-2">
          <SidebarSorter
            sortFn={sortFn}
            setSortFn={setSortFn}
            sortOptions={sortOptions}
            isMobile={isMobile}
          />
        </li>
      </ul>
      <div className="flex-auto">
        <GroupList
          className="p-2"
          groups={sortedGroups}
          pinnedGroups={sortedPinnedGroups}
        />
      </div>
    </nav>
  );
}
