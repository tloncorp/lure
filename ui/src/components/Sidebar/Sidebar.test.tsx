/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { describe, expect, it } from 'vitest';
import { createMockGroup } from '@/mocks/groups';
import { Group } from '@/types/groups';
import { render } from '../../../test/utils';
import Sidebar from './Sidebar';

const fakeFlag = '~zod/tlon';
const fakeGroup: Group = createMockGroup('Fake Group');

vi.mock('@/state/chat', () => ({
  useBriefs: () => ({}),
  usePinnedGroups: () => ({}),
  usePinned: () => [],
  useGetLatestChat: () => () => 0,
  useGetLatestCurio: () => () => 0,
  useGetLatestNote: () => () => 0,
}));

vi.mock('@/state/groups', () => ({
  useGroup: () => fakeGroup,
  useRouteGroup: () => fakeFlag,
  useGroupsInitialized: () => true,
  useGroups: () => [fakeFlag, fakeGroup],
  usePendingInvites: () => [],
}));

vi.mock('@/logic/utils', () => ({
  createStorageKey: () => 'fake-key',
  randomIntInRange: () => 100,
  normalizeUrbitColor: () => '#ffffff',
  hasKeys: () => false,
  randomElement: (a: any[]) => a[0],
}));

describe('Sidebar', () => {
  it('renders as expected', () => {
    const { asFragment } = render(<Sidebar />);
    expect(asFragment()).toMatchSnapshot();
  });
});
