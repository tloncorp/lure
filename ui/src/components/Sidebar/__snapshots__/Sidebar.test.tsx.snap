// Vitest Snapshot v1

exports[`Sidebar > renders as expected 1`] = `
<DocumentFragment>
  <nav
    class="flex h-full w-64 flex-col bg-white"
  >
    <ul
      class="flex w-full flex-col space-y-1 px-2 pt-2"
    >
      <div
        class="group relative flex w-full items-center justify-between rounded-lg text-lg font-semibold sm:text-base text-gray-600 hover:bg-gray-50"
      >
        <button
          class="default-focus flex w-full flex-1 items-center space-x-3 rounded-lg p-2 font-semibold text-black hover:text-gray-800 group"
        >
          <div
            aria-expanded="false"
            aria-haspopup="menu"
            class="appearance-none h-6 w-6 rounded group-hover:bg-gray-100"
            data-state="closed"
            id="radix-0"
            type="button"
          >
            <svg
              class="h-6 w-6 group-hover:hidden"
              fill="none"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="fill-current"
                d="M11.0833 17.9583V16.5833C11.0833 16.0771 11.4937 15.6667 12 15.6667C12.5063 15.6667 12.9167 16.0771 12.9167 16.5833V17.9583C12.9167 20.7428 15.1739 23 17.9583 23C20.7428 23 23 20.7428 23 17.9583C23 15.1739 20.7428 12.9167 17.9583 12.9167H16.5833V13.8333C16.5833 14.3396 16.1729 14.75 15.6667 14.75C15.1604 14.75 14.75 14.3396 14.75 13.8333V10.1667C14.75 9.66041 15.1604 9.25 15.6667 9.25C16.1729 9.25 16.5833 9.66041 16.5833 10.1667V11.0833H17.9583C20.7428 11.0833 23 8.8261 23 6.04167C23 3.25723 20.7428 1 17.9583 1C15.1739 1 12.9167 3.25723 12.9167 6.04167V7.41667C12.9167 7.92293 12.5063 8.33333 12 8.33333C11.4937 8.33333 11.0833 7.92293 11.0833 7.41667V6.04167C11.0833 3.25723 8.8261 1 6.04167 1C3.25723 1 1 3.25723 1 6.04167C1 8.8261 3.25723 11.0833 6.04167 11.0833H7.41667V10.1667C7.41667 9.66041 7.82707 9.25 8.33333 9.25C8.83959 9.25 9.25 9.66041 9.25 10.1667V13.8333C9.25 14.3396 8.83959 14.75 8.33333 14.75C7.82707 14.75 7.41667 14.3396 7.41667 13.8333V12.9167H6.04167C3.25723 12.9167 1 15.1739 1 17.9583C1 20.7428 3.25723 23 6.04167 23C8.8261 23 11.0833 20.7428 11.0833 17.9583Z"
              />
            </svg>
            <svg
              class="m-1 h-4 w-4 text-gray-800 hidden group-hover:block"
              fill="none"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="fill-current"
                clip-rule="evenodd"
                d="M5 6c0-.55228.44772-1 1-1h12c.5523 0 1 .44772 1 1s-.4477 1-1 1H6c-.55228 0-1-.44772-1-1Zm0 6c0-.5523.44772-1 1-1h12c.5523 0 1 .4477 1 1s-.4477 1-1 1H6c-.55228 0-1-.4477-1-1Zm1 5c-.55228 0-1 .4477-1 1s.44772 1 1 1h12c.5523 0 1-.4477 1-1s-.4477-1-1-1H6Z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
          <div
            class="max-w-full flex-1 text-left truncate"
          >
            <div
              class="flex items-center justify-between"
            >
              Groups
              <a
                aria-label="Back to Landscape"
                class="h-6 w-6 no-underline hidden group-hover:block"
                href="/apps/grid"
                rel="noreferrer"
                target="_blank"
                title="Back to Landscape"
              >
                <svg
                  class="text-gray-400"
                  fill="none"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    class="stroke-current"
                    d="M8 14V8M8 8H14M8 8L15 15"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  />
                </svg>
              </a>
            </div>
          </div>
        </button>
      </div>
      <div
        class="h-5"
      />
      <li
        class="group relative flex w-full items-center justify-between rounded-lg text-lg font-semibold sm:text-base text-gray-600"
      >
        <a
          class="default-focus flex w-full flex-1 items-center space-x-3 rounded-lg p-2 font-semibold"
          href="/profile/edit"
        >
          <div
            class="relative flex flex-none items-center justify-center rounded bg-black w-6 h-6 rounded p-1.5"
            style="background-color: rgb(204, 204, 204);"
          >
            <svg
              class=""
              style="display: block; width: 12px; height: 12px;"
              version="1.1"
              viewBox="0 0 12 12"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="hsla(0, 0%, 80%, 1)"
                height="12px"
                width="12px"
                x="0px"
                y="0px"
              />
              <g
                transform="matrix(0.046875,0,0,0.046875,0,0)"
              >
                <g
                  fill="none"
                  transform="matrix(1,0,0,1,0,0)"
                >
                  <rect
                    dataisgeon="true"
                    fill="white"
                    height="128"
                    stroke="hsla(0, 0%, 80%, 1)"
                    stroke-linecap="square"
                    stroke-width="0.8px"
                    vector-effect="non-scaling-stroke"
                    width="128"
                  />
                </g>
                <g
                  fill="none"
                  transform="matrix(1,0,0,1,128,0)"
                >
                  <circle
                    cx="64"
                    cy="64"
                    dataisgeon="true"
                    fill="white"
                    r="64"
                    stroke="hsla(0, 0%, 80%, 1)"
                    stroke-linecap="square"
                    stroke-width="0.8px"
                    vector-effect="non-scaling-stroke"
                  />
                </g>
                <g
                  fill="none"
                  transform="matrix(1,0,0,1,0,128)"
                >
                  <path
                    d="M0.0541 0C70.7217 0.0292317 128 57.3256 128 128C57.3177 128 0.0164917 70.7089 7.62806e-06 0.0305091C7.62851e-06 0.0203397 -4.44317e-10 0.01017 0 0H0.0541Z"
                    dataisgeon="true"
                    fill="white"
                    stroke="hsla(0, 0%, 80%, 1)"
                    stroke-linecap="square"
                    stroke-width="0.8px"
                    vector-effect="non-scaling-stroke"
                  />
                </g>
                <g
                  fill="none"
                  transform="matrix(1,0,0,1,128,128)"
                >
                  <path
                    d="M64 128H0L5.59506e-06 0L64 5.59506e-06C99.3462 8.68512e-06 128 28.6538 128 64C128 99.3462 99.3462 128 64 128Z"
                    dataisgeon="true"
                    fill="white"
                    stroke="hsla(0, 0%, 80%, 1)"
                    stroke-linecap="square"
                    stroke-width="0.8px"
                    vector-effect="non-scaling-stroke"
                  />
                </g>
              </g>
            </svg>
          </div>
          <div
            class="max-w-full flex-1 text-left truncate"
          >
            <span>
              <span
                aria-hidden="true"
              >
                ~
              </span>
              <span>
                finned
              </span>
              <span
                aria-hidden="true"
              >
                -
              </span>
              <span
                aria-hidden="false"
              >
                palmer
              </span>
            </span>
          </div>
        </a>
      </li>
      <li
        class="group relative flex w-full items-center justify-between rounded-lg text-lg font-semibold sm:text-base text-gray-600 bg-gray-100"
      >
        <a
          class="default-focus flex w-full flex-1 items-center space-x-3 rounded-lg p-2 font-semibold"
          href="/notifications"
        >
          <div
            class="flex h-6 w-6 items-center justify-center rounded text-sm font-semibold bg-gray-100"
          >
            <svg
              class="h-6 w-6"
              fill="none"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                class="fill-current"
                height="8"
                rx="4"
                width="8"
                x="8"
                y="8"
              />
            </svg>
          </div>
          <div
            class="max-w-full flex-1 text-left truncate"
            title="Notifications"
          >
            Notifications
          </div>
        </a>
      </li>
      <li
        class="group relative flex w-full items-center justify-between rounded-lg text-lg font-semibold sm:text-base text-gray-600 hover:bg-gray-50"
      >
        <a
          class="default-focus flex w-full flex-1 items-center space-x-3 rounded-lg p-2 font-semibold"
          href="/find"
        >
          <svg
            class="m-1 h-4 w-4"
            fill="none"
            viewBox="0 0 16 16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              class="fill-current"
              clip-rule="evenodd"
              d="M6.00003 7c0-1.65685 1.34314-3 3-3C10.6569 4 12 5.34315 12 7s-1.3431 3-2.99997 3c-1.65686 0-3-1.34315-3-3Zm3-5c-2.76142 0-5 2.23858-5 5 0 1.01907.30487 1.96694.82838 2.75737l-.53552.53553-2 2c-.39052.3905-.39052 1.0237 0 1.4142.39053.3905 1.02369.3905 1.41422 0l2-2 .53551-.5355c.79044.5235 1.73832.8284 2.75741.8284C11.7615 12 14 9.76142 14 7s-2.2385-5-4.99997-5Z"
              fill-rule="evenodd"
            />
          </svg>
          <div
            class="max-w-full flex-1 text-left truncate"
          >
            <div
              class="flex items-center"
            >
              Find Groups
            </div>
          </div>
        </a>
      </li>
      <li
        class="group relative flex w-full items-center justify-between rounded-lg text-lg font-semibold sm:text-base text-gray-600 hover:bg-gray-50"
      >
        <a
          class="default-focus flex w-full flex-1 items-center space-x-3 rounded-lg p-2 font-semibold"
          href="/groups/new"
        >
          <svg
            class="m-1 h-4 w-4"
            fill="none"
            viewBox="0 0 16 16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              class="stroke-current"
              d="M3 8h10M8 3v10"
              stroke-linecap="round"
              stroke-width="2"
            />
          </svg>
          <div
            class="max-w-full flex-1 text-left truncate"
            title="Create Group"
          >
            Create Group
          </div>
        </a>
      </li>
    </ul>
    <ul
      class="flex-initial overflow-y-auto overflow-x-hidden px-2"
    >
      <li
        class="-mx-2 mt-5 grow border-t-2 border-gray-50 pt-3 pb-2"
      >
        <span
          class="ml-4 text-sm font-semibold text-gray-400"
        >
          All Groups
        </span>
      </li>
      <li
        class="relative p-2"
      >
        <button
          aria-expanded="false"
          aria-haspopup="menu"
          aria-label="Groups Sort Options"
          class="default-focus flex w-full items-center justify-between rounded-lg bg-gray-50 py-1 px-2 text-sm font-semibold"
          data-state="closed"
          id="radix-1"
          type="button"
        >
          <span
            class="flex items-center"
          >
            <svg
              class="h-4 w-4 text-gray-400"
              fill="none"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="fill-current"
                clip-rule="evenodd"
                d="M17.207 4.793a1 1 0 0 0-1.414 0l-3 3a1 1 0 0 0 1.414 1.414L15.5 7.914V17.5a1 1 0 1 0 2 0V7.914l1.293 1.293a1 1 0 1 0 1.414-1.414l-3-3ZM8.5 5.5a1 1 0 0 0-2 0v9.586l-1.293-1.293a1 1 0 0 0-1.414 1.414l3 3a1 1 0 0 0 1.414 0l3-3a1 1 0 0 0-1.414-1.414L8.5 15.086V5.5Z"
                fill-rule="evenodd"
              />
            </svg>
            <span
              class="mr-2 pl-1"
            >
              Sort: A → Z
            </span>
          </span>
          <svg
            class="h-4 w-4 text-gray-400"
            fill="none"
            viewBox="0 0 16 16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              class="stroke-current"
              d="m4 6 4 4 4-4"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </button>
      </li>
    </ul>
    <div
      class="flex-auto"
    >
      <ul
        class="h-full animate-pulse space-y-3 p-2 sm:space-y-0"
      >
        <li
          class="relative flex w-full items-center justify-between rounded-lg"
        >
          <div
            class="flex w-full flex-1 items-center space-x-3 rounded-lg p-2"
          >
            <div
              class="flex h-12 w-12 items-center justify-center rounded-md bg-gray-100 text-sm sm:h-6 sm:w-6"
            />
            <div
              class="bg-gray-100 h-4 rounded-md"
              style="width: 100px;"
            />
          </div>
        </li>
        <li
          class="relative flex w-full items-center justify-between rounded-lg"
        >
          <div
            class="flex w-full flex-1 items-center space-x-3 rounded-lg p-2"
          >
            <div
              class="flex h-12 w-12 items-center justify-center rounded-md bg-gray-100 text-sm sm:h-6 sm:w-6"
            />
            <div
              class="bg-gray-100 h-4 rounded-md"
              style="width: 100px;"
            />
          </div>
        </li>
      </ul>
    </div>
  </nav>
</DocumentFragment>
`;
